### ISSUE TITLE
Inconsistent conditional type results.

### ISSUE THREAD

**Issue by Lotarcc** [hashicorp/terraform#28293]
**Terraform version**
`Terraform v0.14.9`

**Terraform Configuration Files**

```terraform
locals  {

  mke_launchpad_tmpl = {
    apiVersion = "api"
    kind       = "mke"
    spec = {
      mke = {
        version: var.mke_version
        adminUsername = "admin"
        adminPassword = var.admin_password
        installFlags : [
          "--default-node-orchestrator=kubernetes",
          "--san=${module.masters.lb_dns_name}",
        ]
        licenseFilePath: var.license_file_path
        upgradeFlags: [
          "--force-minimums",
          "--force-recent-backups",
        ]
      }
      mcr = {
        version: var.mcr_version
      }
      msr = {}
      hosts = concat(local.managers, local.workers, local.windows_workers)
    }
  }


  msr_launchpad_tmpl = {
    apiVersion = "api"
    kind       = "mke+msr"
    spec = {
      mke = {
        version: var.mke_version
        adminUsername = "admin"
        adminPassword = var.admin_password
        installFlags : [
          "--default-node-orchestrator=kubernetes",
          "--san=${module.masters.lb_dns_name}",
        ]
        licenseFilePath: var.license_file_path
        upgradeFlags: [
          "--force-minimums",
          "--force-recent-backups",
        ]
      }
      mcr = {
        version: var.mcr_version
      }
      msr = {
        version: var.msr_version
        installFlags : [
          "--ucp-insecure-tls",
          "--dtr-external-url ${module.msrs.lb_dns_name}",
        ]
      }
      hosts = concat(local.managers, local.msrs, local.workers, local.windows_workers)
    }
  }

  launchpad_tmpl = var.msr_count > 0 ? local.msr_launchpad_tmpl : local.mke_launchpad_tmpl
}
```

***Expected behavior***

It will pass `msr_version` to the template and do terraform plan or apply without dropping.

***Actual behavior***
Without `msr_version` variable it works fine, but with it - crashing.
```
Error: Inconsistent conditional result types

  on main.tf line 179, in locals:
 179:   launchpad_tmpl = var.msr_count > 0 ? local.msr_launchpad_tmpl : local.mke_launchpad_tmpl
    |----------------
    | local.mke_launchpad_tmpl is object with 3 attributes
    | local.msr_launchpad_tmpl is object with 3 attributes
    | var.msr_count is 1

The true and false result expressions must have consistent types. The given
expressions are object and object, respectively.
```

**Comment by jbardin** [comment]
Hi @Lotarcc 

Thanks for filing the issue. The example here is not complete enough to reproduce the error, but there have been improvements to the evaluation of conditional expressions like this in 0.15. Is it possible to replicate this problem with the latest 0.15 release? (currently 0.15.0-rc1)

If you can still reproduce the issue in 0.15, please try to create a minimal reproducible example that demonstrates the issue.

Thanks!

**Comment by Lotarcc** [comment]
HI @jbardin

Thank you for a quick response.
In order to try to reproduce it on version 0.15 I will need to rewrite a lot of code cause of `map` deprecation.

**Comment by Lotarcc** [comment]
@jabrdin Maybe you can do a quick look in this code since the issue is hiding only there?

**Comment by jbardin** [comment]
@Lotarcc, I have not tried to put together a reproduction if this myself yet, as there are a number of undefined values in the config. From a visual inspection, I am guessing that the `spec` objects are were the real difference lies, as the `msr` values will have different types. These might be resolvable in 0.15, but I wouldn't be surprised if they are not yet as I don't recall any changes to the unification process for nested objects specifically.

If the `msr` attribute type is the issue here, one way to handle this might be to use a module with input variables to define strict types for the pieces of your spec and return them as an output.

**Comment by apparentlymart** [comment]
Hi @Lotarcc,

I think I found and answered [your question on Stack Overflow](https://stackoverflow.com/questions/66970542/inconsistent-conditional-result-types-with-locals-templatization) already, because I found that before I saw this issue.

Based on what I saw over there it seemed to me like Terraform was behaving as intended, aside from this error message not being particularly helpful due to Terraform not knowing how to talk about a type mismatch deep inside two object types. Hopefully one of the two approaches I suggested in my Stack Overflow answer will get you a working solution.
